libcudacxx/include/cuda/std/inplace_vector

```diff
@@ -0,0 +1,1746 @@
+{
+  using size_type      = size_t;
+  using __size_type    = __inplace_vector_size_type<_Capacity>;
+  using reference      = _Tp&;
+  using iterator       = _Tp*;
+  using const_iterator = const _Tp*;
+
+  _Tp __elems_[_Capacity] = {};
+  __size_type __size_{0};
+
+  // [containers.sequences.inplace.vector.cons], construct/copy/destroy
+  constexpr __inplace_vector_base()                                        = default;
+  constexpr __inplace_vector_base(const __inplace_vector_base&)            = default;
+  constexpr __inplace_vector_base& operator=(const __inplace_vector_base&) = default;
+
+  _LIBCUDACXX_INLINE_VISIBILITY constexpr __inplace_vector_base(__inplace_vector_base&& __other) noexcept
+      : __elems_()
+      , __size_(__other.__size_)
+  {
+    _CUDA_VSTD::copy(__other.__elems_, __other.__elems_ + __other.__size_, __elems_);
+    __other.__size_ = 0;
+  }
+
+  _LIBCUDACXX_INLINE_VISIBILITY constexpr __inplace_vector_base& operator=(__inplace_vector_base&& __other) noexcept
+  {
+    _CUDA_VSTD::copy(__other.__elems_, __other.__elems_ + __other.__size_, __elems_);
+    __size_         = __other.__size_;
+    __other.__size_ = 0;
+    return *this;
+  }
+
+  // [containers.sequences.inplace.vector.members] size/capacity
+  _LIBCUDACXX_INLINE_VISIBILITY constexpr void __change_size(const ptrdiff_t __count) noexcept
+  {
+    __size_ += __count;
+  }
+
+  _CCCL_NODISCARD _LIBCUDACXX_INLINE_VISIBILITY constexpr size_type size() const noexcept
+  {
+    return static_cast<size_type>(__size_);
+  }
+
+  _CCCL_NODISCARD _LIBCUDACXX_INLINE_VISIBILITY constexpr bool empty() const noexcept
+  {
+    return __size_ == 0;
+  }
+
+  // [containers.sequences.inplace.vector.data], data access
+  _CCCL_NODISCARD _LIBCUDACXX_INLINE_VISIBILITY constexpr _Tp* data() noexcept
+  {
+    return __elems_;
+  }
+
+  _CCCL_NODISCARD _LIBCUDACXX_INLINE_VISIBILITY constexpr const _Tp* data() const noexcept
+  {
+    return __elems_;
+  }
+
+  // [containers.sequences.inplace.vector.itertators] iterators
+  _CCCL_NODISCARD _LIBCUDACXX_INLINE_VISIBILITY constexpr iterator begin() noexcept
+  {
+    return __elems_;
+  }
+
+  _CCCL_NODISCARD _LIBCUDACXX_INLINE_VISIBILITY constexpr const_iterator begin() const noexcept
+  {
+    return __elems_;
+  }
+
+  _CCCL_NODISCARD _LIBCUDACXX_INLINE_VISIBILITY constexpr iterator end() noexcept
+  {
+    return __elems_ + __size_;
+  }
+
+  _CCCL_NODISCARD _LIBCUDACXX_INLINE_VISIBILITY constexpr const_iterator end() const noexcept
+  {
+    return __elems_ + __size_;
+  }
+
+  // [containers.sequences.inplace.vector.modifiers], modifiers
+  template <class... _Args>
+  _LIBCUDACXX_INLINE_VISIBILITY constexpr reference
+  unchecked_emplace_back(_Args&&... __args) noexcept(_CCCL_TRAIT(is_nothrow_constructible, _Tp, _Args...))
+  {
+    _Tp* __final = __elems_ + __size_;
+    *__final     = _Tp(_CUDA_VSTD::forward<_Args>(__args)...);
+    ++__size_;
+    return *__final;
+  }
```

Comment:
Question: Why not placement new? Like: `new(__final) _Tp(_CUDA_VSTD::forward<_Args>(__args)...);`

Reply:
This is the `is_trivial` specialization, which is required to be constexpr. That means we must avoid placement new

Reply:
Ah, I see! So we have to wait until C++26 for placement new to be allowed in a constant evaluation: https://wg21.link/p2747